pkg/kubecost/allocation.go

```diff
@@ -2260,8 +2260,39 @@ func (asr *AllocationSetRange) Accumulate() (*AllocationSet, error) {
 	return allocSet, nil
 }
 
-// TODO accumulate into lower-resolution chunks of the given resolution
-// func (asr *AllocationSetRange) AccumulateBy(resolution time.Duration) *AllocationSetRange
+// AccumulateBy sums AllocationSets based on the resolution given. The resolution given is subject to the scale used for the AllocationSets.
+// If the requested resolution is smaller than the window of an AllocationSet then the resolution will default to the duration of a set.
+// Resolutions larger than the duration of the entire AllocationSetRange will default to the duration of the range.
+func (asr *AllocationSetRange) AccumulateBy(resolution time.Duration) (*AllocationSetRange, error) {
+	allocSetRange := &AllocationSetRange{allocations: []*AllocationSet{}}
+	var allocSet *AllocationSet
+	var currAccumulatedSum time.Duration
+	var err error
+
+	asr.Lock()
+	defer asr.Unlock()
+
+	for i, as := range asr.allocations {
+		allocSet, err = allocSet.accumulate(as)
+		if err != nil {
+			return nil, err
+		}
+
+		if allocSet != nil {
+			currAccumulatedSum += allocSet.Window.Duration()
+
+			// check if end of asr to sum the final set
+			// If accumulated sum !>= resolution return 1 accumulated set
+			if currAccumulatedSum >= resolution || i == len(asr.allocations)-1 {
+				allocSetRange.allocations = append(allocSetRange.allocations, allocSet)
+				allocSet = &AllocationSet{allocations: map[string]*Allocation{}}
```

Comment:
Same, RE using the existing constructor function. And here, there's actually a very good reason! There are other structures that `NewAllocationSet` instantiates, which we'd need to handle. Otherwise, this might nil panic.

Reply:
Makes sense. The initial concern was that initiating a new AllocationSet would result in an incorrect accumulation based on the window start/end parameters we're forced to include in the call. However I verified that AllocationSet's IsEmpty() will still return empty with non-nil window durations when accumulating so fixing this shouldn't cause issues

Fixed!